### PR TITLE
Add DLQ Admin APIs and Integrate with CLI

### PR DESCRIPTION
**Pull request by yux0**
This PR contains:
1. Domain DLQ implementation
2. Admin APIs
3. CLI for DLQ operations

Operations:
1. Read DLQ messages by Max message ID or number of messages
2. Purge DLQ messages by Max message ID or purge all
3. Merge DLQ messages by Max message ID or merge all

### CODE REVIEW

File: service/frontend/workflowHandler.go [hunk 1]
```diff
@@ -115,6 +115,7 @@ var (
 	errQueryDisallowedForDomain                   = &gen.BadRequestError{Message: "Domain is not allowed to query, please contact cadence team to re-enable queries."}
 	errClusterNameNotSet                          = &gen.BadRequestError{Message: "Cluster name is not set."}
 	errEmptyReplicationInfo                       = &gen.BadRequestError{Message: "Replication task info is not set."}
+	errEmptyQueueType                             = &gen.BadRequestError{Message: "Queue type is not set."}
```
Comment (yycptt) [inline service/frontend/workflowHandler.go]: Is this error only used in adminHandler?

File: service/frontend/adminHandler.go [hunk 2]
```diff
@@ -87,6 +89,11 @@ func NewAdminHandler(
 		numberOfHistoryShards: params.PersistenceConfig.NumHistoryShards,
 		params:                params,
 		config:                config,
+		domainDLQHandler: domain.NewDLQTaskHandler(
+			resource.GetDomainReplicationTaskHandler(),
+			resource.GetDomainReplicationQueue(),
+			resource.GetLogger().WithTags(),
```
Comment (yycptt) [inline service/frontend/adminHandler.go]: `WithTags()`?

File: service/frontend/adminHandler.go [hunk 3]
```diff
@@ -29,8 +29,8 @@ import (
 	"time"
 
 	"github.com/olivere/elastic"
-
 	"github.com/pborman/uuid"
+	"go.uber.org/cadence/.gen/go/shared"
```
Comment (yycptt) [inline service/frontend/adminHandler.go]: It's already imported under the name `gen`

File: common/persistence/cassandra/cassandraQueue.go [hunk 4]
```diff
@@ -364,16 +382,62 @@ func (q *cassandraQueue) UpdateAckLevel(
...
+		}
+	}
+
+	// Ignore possibly delayed message
+	if queueMetadata.clusterAckLevels[clusterName] > messageID {
+		return nil
+	}
+
+	queueMetadata.clusterAckLevels[clusterName] = messageID
+	queueMetadata.version++
```
Comment (vancexu) [inline common/persistence/cassandra/cassandraQueue.go]: If this should always be done together, consider making it a function and queueMetadata

File: common/persistence/domainReplicationQueue.go [hunk 5]
```diff
@@ -169,6 +180,76 @@ func (q *domainReplicationQueueImpl) GetAckLevels() (map[string]int, error) {
 	return q.queue.GetAckLevels()
 }
 
+func (q *domainReplicationQueueImpl) GetMessagesFromDLQ(
```
Comment (vancexu) [inline common/persistence/domainReplicationQueue.go]: can we add unit test for these new added functions?

File: service/frontend/adminHandler.go [hunk 6]
```diff
@@ -776,6 +783,132 @@ func (adh *AdminHandler) ReapplyEvents(
...
+		return adh.error(errEmptyQueueType, scope)
+	}
+
+	if !request.IsSetInclusiveEndMessageID() {
+		request.InclusiveEndMessageID = common.Int64Ptr(common.EndMessageID)
+	}
+
+	switch request.GetQueueType() {
+	case admin.QueueTypeReplication:
+		return &shared.InternalServiceError{Message: "Not implement."}
```
Comment (vancexu) [inline service/frontend/adminHandler.go]: BadRequestError?

File: service/frontend/adminHandler.go [hunk 7]
```diff
@@ -776,6 +783,132 @@ func (adh *AdminHandler) ReapplyEvents(
...
+
+	if !request.IsSetQueueType() {
+		return adh.error(errEmptyQueueType, scope)
+	}
+
+	if !request.IsSetInclusiveEndMessageID() {
+		request.InclusiveEndMessageID = common.Int64Ptr(common.EndMessageID)
+	}
+
+	switch request.GetQueueType() {
```
Comment (vancexu) [inline service/frontend/adminHandler.go]: add a default just for safety

Comment (vancexu) [reply service/frontend/adminHandler.go]: same for the other two

File: service/frontend/adminHandler.go (hunk 6 above)
Comment (vancexu) [reply service/frontend/adminHandler.go]: same for the other two

File: tools/cli/flags.go [hunk 8]
```diff
@@ -208,6 +208,12 @@ const (
 	FlagTLSKeyPath                        = "tls_key_path"
 	FlagTLSCaPath                         = "tls_ca_path"
 	FlagTLSEnableHostVerification         = "tls_enable_host_verification"
+	FlagDLQType                           = "dlq_type"
+	FlagDLQTypeWithAlias                  = FlagDLQType + ", dt"
+	FlagMaxMessageCount                   = "max_message_size"
```
Comment (vancexu) [inline tools/cli/flags.go]: consistently using "max_message_count"?

File: common/domain/dlqTaskHandler.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	"github.com/uber/cadence/.gen/go/replicator"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/persistence"
+)
+
+type (
+	// DLQTaskHandler is the interface handles domain DLQ messages
+	DLQTaskHandler interface {
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/domain/dlqTaskHandler.go]: Updated the API names

File: common/domain/replicationTaskHandler.go [hunk 10]
```diff
@@ -53,45 +55,49 @@ var (
 // NOTE: the counterpart of domain replication transmission logic is in service/fropntend package
 
 type (
-	// DomainReplicator is the interface which can replicate the domain
-	DomainReplicator interface {
+	// ReplicationHandler is the interface which can replicate the domain
+	ReplicationHandler interface {
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/domain/replicationTaskHandler.go]: This only handles the replication task. The replicator may have other meaning

File: tools/cli/adminDLQCommands.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+		task := item.(*replicator.ReplicationTask)
+		taskStr, err := json.Marshal(task)
+		if err != nil {
+			ErrorAndExit(fmt.Sprintf("fail to envode dlq message. Last read message id: %v", lastReadMessageID), err)
+		}
+
+		lastReadMessageID = int(*task.SourceTaskId)
+		maxMessageCount--
+		_, err = outputFile.WriteString(fmt.Sprintf("%v\n", string(taskStr)))
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/adminDLQCommands.go]: Log and exit if there is an error.

File: tools/cli/adminDLQCommands.go [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+func AdminGetDLQInfo(c *cli.Context) {
+	ctx, cancel := newContext(c)
+	defer cancel()
+
+	adminClient := cFactory.ServerAdminClient(c)
+	dlqType := getRequiredOption(c, FlagDLQType)
+	outputFile := getOutputFile(c.String(FlagOutputFilename))
+	defer outputFile.Close()
+
+	maxMessageCount := common.EndMessageID
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/adminDLQCommands.go]: Updated

File: tools/cli/adminDLQCommands.go [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+	iterator := collection.NewPagingIterator(paginationFunc)
+	var lastReadMessageID int
+	for iterator.HasNext() && maxMessageCount > 0 {
+		item, err := iterator.Next()
+		if err != nil {
+			ErrorAndExit(fmt.Sprintf("fail to read dlq message. Last read message id: %v", lastReadMessageID), err)
+		}
+
+		task := item.(*replicator.ReplicationTask)
+		taskStr, err := json.Marshal(task)
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/adminDLQCommands.go]: Updated

File: tools/cli/adminDLQCommands.go [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+		}
+	}
+
+	adminClient := cFactory.ServerAdminClient(c)
+	if err := adminClient.PurgeDLQMessages(ctx, &admin.PurgeDLQMessagesRequest{
+		QueueType:             toQueueType(dlqType),
+		InclusiveEndMessageID: lastMessageID,
+	}); err != nil {
+		ErrorAndExit("Failed to purge dlq", nil)
+	}
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/adminDLQCommands.go]: Updated

File: tools/cli/adminDLQCommands.go [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+	} else {
+		fmt.Println("Are you sure to merge all DLQ messages without a upper boundary? (Y/n)")
+		reader := bufio.NewReader(os.Stdin)
+		confirm, err := reader.ReadByte()
+		if err != nil {
+			panic(err)
+		}
+		if confirm != 'Y' {
+			osExit(0)
+		}
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/adminDLQCommands.go]: Updated

File: tools/cli/adminDLQCommands.go [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+		return adminClient.MergeDLQMessages(ctx, &admin.MergeDLQMessagesRequest{
+			QueueType:             toQueueType(dlqType),
+			InclusiveEndMessageID: lastMessageID,
+			MaximumPageSize:       common.Int32Ptr(defaultPageSize),
+			NextPageToken:         token,
+		})
+	}
+
+	var pageToken []byte
+	for hasMore := true; hasMore; hasMore = len(pageToken) > 0 {
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/adminDLQCommands.go]: Updated

File: tools/cli/adminDLQCommands.go [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+func toQueueType(dlqType string) *admin.QueueType {
+	switch dlqType {
+	case "domain":
+		return admin.QueueTypeDomain.Ptr()
+	case "history":
+		return admin.QueueTypeReplication.Ptr()
+	default:
+		ErrorAndExit("The queue type is not supported.", fmt.Errorf("the queue type is not supported. Type: %v", dlqType))
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/adminDLQCommands.go]: Yes. Just to make complier happy

File: tools/cli/admin.go [hunk 18]
```diff
@@ -795,3 +795,82 @@ func newAdminClusterCommands() []cli.Command {
 		},
 	}
 }
+
+func newAdminDLQCommands() []cli.Command {
+	return []cli.Command{
+		{
+			Name:    "read",
+			Aliases: []string{"i"},
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/admin.go]: Updated

File: tools/cli/admin.go [hunk 19]
```diff
@@ -795,3 +795,82 @@ func newAdminClusterCommands() []cli.Command {
...
+			Usage:   "Read DLQ Messages",
+			Flags: []cli.Flag{
+				cli.StringFlag{
+					Name:  FlagDLQTypeWithAlias,
+					Usage: "Type of DLQ to manage",
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/admin.go]: Updated with the available options

File: tools/cli/adminDLQCommands.go [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/collection"
+)
+
+const (
+	defaultPageSize = 1000
+)
+
+// AdminGetDLQInfo get DLQ metadata
+func AdminGetDLQInfo(c *cli.Context) {
```
[the remark this answers is not in the document]
Comment (yux0) [reply tools/cli/adminDLQCommands.go]: Updated

File: service/frontend/workflowHandler.go (hunk 1 above)
Comment (yux0) [reply service/frontend/workflowHandler.go]: Yes

File: service/frontend/adminHandler.go (hunk 2 above)
Comment (yux0) [reply service/frontend/adminHandler.go]: Removed

File: common/persistence/cassandra/cassandraQueue.go [hunk 21]
```diff
@@ -364,16 +382,62 @@ func (q *cassandraQueue) UpdateAckLevel(
...
 
-func (q *cassandraQueue) getQueueMetadata() (*queueMetadata, error) {
-	query := q.session.Query(templateGetQueueMetadataQuery, q.queueType)
+func (q *cassandraQueue) UpdateDLQAckLevel(
+	messageID int,
+	clusterName string,
+) error {
+
+	// Use negative queue type as the dlq type
+	queueMetadata, err := q.getQueueMetadata(-q.queueType)
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/persistence/cassandra/cassandraQueue.go]: Updated

File: common/persistence/cassandra/cassandraQueue.go [hunk 22]
```diff
@@ -395,12 +459,13 @@ func (q *cassandraQueue) getQueueMetadata() (*queueMetadata, error) {
 
 func (q *cassandraQueue) updateQueueMetadata(
 	metadata *queueMetadata,
+	queueType common.QueueType,
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/persistence/cassandra/cassandraQueue.go]: metadata currently uses to store the levels. It stores by different queue type

File: common/persistence/cassandra/cassandraQueue.go (hunk 4 above)
Comment (yux0) [reply common/persistence/cassandra/cassandraQueue.go]: Moved to a common func

File: common/persistence/domainReplicationQueue.go (hunk 5 above)
Comment (yux0) [reply common/persistence/domainReplicationQueue.go]: It is a little hard to add the unit test b/c the queue interface mock generation. I will try to add one

File: service/frontend/adminHandler.go (hunk 6 above)
Comment (yux0) [reply service/frontend/adminHandler.go]: Updated

File: service/frontend/adminHandler.go (hunk 7 above)
Comment (yux0) [reply service/frontend/adminHandler.go]: Updated

File: tools/cli/flags.go (hunk 8 above)
Comment (yux0) [reply tools/cli/flags.go]: Updated

File: common/domain/dlqMessageHandler.go [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+			return nil, &shared.InternalServiceError{Message: "Encounter non domain replication task in domain replication queue."}
+		}
+
+		if err := d.replicationHandler.HandleReceivingTask(
+			domainTask,
+		); err != nil {
+			return nil, err
+		}
+
+		if err := d.domainReplicationQueue.DeleteMessageFromDLQ(int(*message.SourceTaskId)); err != nil {
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/domain/dlqMessageHandler.go]: Updated

File: common/domain/transmissionTaskHandler_test.go [hunk 24]
```diff
@@ -18,6 +18,8 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
+//+build test
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/domain/transmissionTaskHandler_test.go]: removed

File: service/worker/replicator/replicator.go [hunk 25]
```diff
@@ -90,6 +91,7 @@ func NewReplicator(
 	hostInfo *membership.HostInfo,
 	serviceResolver membership.ServiceResolver,
 	domainReplicationQueue persistence.DomainReplicationQueue,
+	domainReplicationTask domain.ReplicationHandler,
```
[the remark this answers is not in the document]
Comment (yux0) [reply service/worker/replicator/replicator.go]: Updated

File: common/resource/resource.go [hunk 26]
```diff
@@ -65,6 +66,7 @@ type (
 		GetMetricsClient() metrics.Client
 		GetArchiverProvider() provider.ArchiverProvider
 		GetMessagingClient() messaging.Client
+		GetDomainReplicationTaskHandler() domain.ReplicationHandler
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/resource/resource.go]: Removed this from the resource

File: common/domain/dlqMessageHandler.go [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+				Message: fmt.Sprintf("Failed to delete DLQ message after merge. Last Merged Message ID: %v, Error: %v.",
+					ackedMessageID,
+					err,
+				),
+			}
+		}
+		ackedMessageID = int(*message.SourceTaskId)
+	}
+
+	if err := d.domainReplicationQueue.UpdateDLQAckLevel(ackedMessageID); err != nil {
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/domain/dlqMessageHandler.go]: Updated.

File: common/domain/replicationTaskHandler.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (yycptt) [reply common/domain/replicationTaskHandler.go]: Please update the file name to match the interface. Or consider sync the name with the other pr, call it `ReplicationTaskExecutor`?

File: service/worker/replicator/replicator.go [hunk 28]
```diff
@@ -48,7 +49,7 @@ type (
 	Replicator struct {
 		domainCache            cache.DomainCache
 		clusterMetadata        cluster.Metadata
-		domainReplicator       DomainReplicator
+		domainReplicator       domain.ReplicationHandler
```
Comment (yycptt) [inline service/worker/replicator/replicator.go]: Also update the field name?

Comment (yux0) [reply service/worker/replicator/replicator.go]: done

File: service/frontend/adminHandler.go [hunk 29]
```diff
@@ -776,6 +792,155 @@ func (adh *AdminHandler) ReapplyEvents(
...
+			tasks, token, err = adh.domainDLQHandler.Read(
+				int(request.GetInclusiveEndMessageID()),
+				int(request.GetMaximumPageSize()),
+				request.GetNextPageToken())
+			return err
+		}
+	default:
+		return nil, &gen.BadRequestError{Message: "The DLQ type is not supported."}
+	}
+	err = backoff.Retry(op, adminServiceRetryPolicy, common.IsServiceTransientError)
```
[the remark this answers is not in the document]
Comment (yux0) [reply service/frontend/adminHandler.go]: done

File: service/worker/replicator/replicator.go [hunk 30]
```diff
@@ -90,23 +91,24 @@ func NewReplicator(
...
-		client:                 client,
-		logger:                 logger,
-		metricsClient:          metricsClient,
-		historySerializer:      persistence.NewPayloadSerializer(),
-		domainReplicationQueue: domainReplicationQueue,
+		hostInfo:                      hostInfo,
+		serviceResolver:               serviceResolver,
+		domainCache:                   domainCache,
+		clusterMetadata:               clusterMetadata,
+		domainReplicationTaskExecutor: domainReplicationTaskExecutor,
```
Comment (yycptt) [inline service/worker/replicator/replicator.go]: `taskExecutor`?

Comment (yux0) [reply service/worker/replicator/replicator.go]: updated

**Pull request merged**